app/services/fees/apply_provider_taxes_service.rb

```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+module Fees
+  class ApplyProviderTaxesService < BaseService
+    def initialize(fee:, fee_taxes:)
+      @fee = fee
+      @fee_taxes = fee_taxes
+
+      super
+    end
+
+    def call
+      result.applied_taxes = []
+      return result if fee.applied_taxes.any?
+
+      applied_taxes_amount_cents = 0
+      applied_taxes_rate = 0
+
+      fee_taxes.tax_breakdown.each do |tax|
+        tax_rate = tax.rate.to_f * 100
+
+        applied_tax = Fee::AppliedTax.new(
+          tax_description: tax.type,
+          tax_code: tax.name.parameterize(separator: '_'),
+          tax_name: tax.name,
+          tax_rate: tax_rate,
+          amount_currency: fee.amount_currency
+        )
+        fee.applied_taxes << applied_tax
+
+        tax_amount_cents = tax.tax_amount
+        applied_tax.amount_cents = tax_amount_cents.round
+        applied_tax.save! if fee.persisted?
+
+        applied_taxes_amount_cents += tax_amount_cents
+        applied_taxes_rate += tax_rate
+
+        result.applied_taxes << applied_tax
+      end
+
+      fee.taxes_amount_cents = applied_taxes_amount_cents.round
```

Comment:
I'm wondering, wouldn't it potentially lead to not matching sum of applied taxes with fee.taxes_amount_cents and if it can be a problem 
(because for applied tax we're rounding each element, but for tax_amount_cents we're rounding the sum)

Reply:
I think both round calls are pretty extra since we should receive amount in cents (and it cents there should be no decimals). I just added it as a guard but we should be good IMO